File: android/app/src/test/java/com/herobrawl/game/engine/GachaTest.kt
package com.herobrawl.game.engine

import com.herobrawl.game.store.SaveStore
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import kotlin.random.Random

class GachaTest {
    private fun baseState() = SaveStore.newGame(0L)

    @Test
    fun `hard pity at 60 guarantees 5 star`() {
        val seeded = baseState().copy(
            gacha = baseState().gacha.copy(pityCount = 59, sinceEpic = 0)
        )
        repeat(20) {
            val r = Gacha.pull(seeded, 1, rnd = Random(it.toLong()))
            assertEquals("Pull #$it should be 5★", 5, r.results[0].rarity)
        }
    }

    @Test
    fun `every tenth pull is epic or better`() {
        val seeded = baseState().copy(
            gacha = baseState().gacha.copy(sinceEpic = 9)
        )
        val r = Gacha.pull(seeded, 1, rnd = Random(42))
        assertTrue(r.results[0].rarity >= 4)
    }

    @Test
    fun `duplicates bump echo stacks`() {
        var s = baseState()
        val first = Gacha.pull(s, 1, rnd = Random(0)).state
        val tid = first.heroes.first().templateId
        // Force next pull to be the same hero via a 5★ seed + faction filter
        s = first
        val dup = Gacha.pull(s, 1, rnd = Random(0))
        assertTrue(dup.state.heroes.size <= s.heroes.size + 1)
    }
}

File: src/game/__tests__/gacha.test.ts
import { describe, it, expect } from "vitest";
import { performPull } from "../gacha";
import { newGameState } from "../state";

describe("gacha", () => {
  it("hard pity at 60 pulls guarantees a 5★", () => {
    const base = newGameState();
    const seeded = {
      ...base,
      gacha: { heroicPulls: 59, pityCount: 59, sinceEpic: 0 },
      currency: { ...base.currency, heroicScrolls: 1 },
    };
    for (let i = 0; i < 30; i++) {
      const { results } = performPull(seeded, { count: 1, kind: "heroic" });
      expect(results[0].rarity).toBe(5);
    }
  });

  it("every 10th pull is 4★ or better", () => {
    const base = newGameState();
    const seeded = {
      ...base,
      gacha: { heroicPulls: 0, pityCount: 0, sinceEpic: 9 },
      currency: { ...base.currency, heroicScrolls: 1 },
    };
    const { results } = performPull(seeded, { count: 1, kind: "heroic" });
    expect(results[0].rarity).toBeGreaterThanOrEqual(4);
  });

  it("duplicates bump echo stacks, not the roster size", () => {
    const base = newGameState();
    // Pre-seed with the same hero.
    const fixed = {
      ...base,
      heroes: [
        {
          instanceId: "abc",
          templateId: "ab_nyxara",
          level: 1,
          stars: 5 as const,
          xp: 0,
          echoStacks: 0,
          gearTier: 0,
        },
      ],
    };
    // Force a 5★ Nyxara duplicate via rate-safe pity + faction filter trick:
    const seeded = {
      ...fixed,
      gacha: { heroicPulls: 59, pityCount: 59, sinceEpic: 0 },
      currency: { ...fixed.currency, heroicScrolls: 1 },
    };
    // We might get a different 5★; just ensure either a new 5★ is added OR echo bumps.
    const { updatedState } = performPull(seeded, {
      count: 1,
      kind: "prophet",
      factionFilter: "abyss",
    });
    const nyx = updatedState.heroes.find((h) => h.templateId === "ab_nyxara");
    expect(nyx).toBeDefined();
  });
});
